fix(scrollbind): properly take filler/virtual lines into account

lewis6991 opened the pull request:
Problem:

`'scrollbind'` does not work properly if the window being scrolled
automatically contains any filler/virtual lines (except for diff filler
lines).

This is because when the scrollbind check is done, the logic only
considers changes to topline which are represented as line numbers.

Solution:

Write the logic for determine the scroll amount to take into account
filler/virtual lines.

Fixes #29751

lewis6991 committed:
fix(scrollbind): properly take filler/virtual lines into account

Problem:

`'scrollbind'` does not work properly if the window being scrolled
automatically contains any filler/virtual lines (except for diff filler
lines).

This is because when the scrollbind check is done, the logic only
considers changes to topline which are represented as line numbers.

Solution:

Write the logic for determine the scroll amount to take into account
filler/virtual lines.

Fixes #29751

lewis6991 commented:
Not super with this solution, but it seems to work.

zeertzjq commented on src/nvim/normal.c:
```diff
@@ -2145,23 +2147,40 @@ void do_check_scrollbind(bool check)
+  int bot = MAX(orig_topline, old_topline);
+  int buflen = orig_wp->w_buffer->b_ml.ml_line_count;
+  for (int i = top + 1; i <= bot; i++) {
+    count += plines_win(orig_wp, i, buflen);
```
Using `plines_win()` may be incorrect here, as that counts the number of screen lines, while `scrollup()` and `scrolldown()` may scroll by screen lines or text lines depending on the value of the 'smoothscroll' option.

Actually, previously (and in Vim) the opposite problem existed, where the number of text lines is always used.

lewis6991 replied:
I haven't put any work in to make this work for smooth scroll, which afaik is still a bit experimental/new.

Either way, using plines should still be better than using text lines as `scrollup()`/`scrolldown()` specifically handle topfill, so we need to send numbers to it which include filler lines.

What I'm not sure about is whether the number we pass should include folds or not.

zeertzjq replied:
I was more talking about wrapped lines, not filler lines.

lewis6991 replied:
I know, the changes here don't improve the situation for wrapped lines, but I don't think regresses it either. Or does it regress because plines respects wrapping? If so I can avoid using plines.

zeertzjq replied:
I think it does regress. When a line is wrapped is one window, it's more common for the corresponding line to also be wrapped (and occupy the same number of screen lines) in the other window. When scrolling over such a line with 'nosmoothscroll', the old behavior works correctly, but the new behavior

zeertzjq commented:
The commit message is missing a 't' at the end

zeertzjq commented on src/nvim/normal.c:
```diff
@@ -2145,23 +2147,39 @@ void do_check_scrollbind(bool check)
 
+static int get_scrollbind_amount(win_T *orig_wp, int old_topline, int topfill_diff)
+{
+  int orig_topline = orig_wp->w_topline;
```
The `orig_topline` name sounds confusing as it's not clear that it's newer than `old_topline`.

zeertzjq commented on src/nvim/normal.c:
```diff
@@ -2145,23 +2147,39 @@ void do_check_scrollbind(bool check)
+    return -topfill_diff;
+  }
+
+  int count = -abs(topfill_diff);
```
This doesn't seems correct. I think it should be like this:
```suggestion
  int count = orig_topline > old_topline ? -topfill_diff : topfill_diff;
```

zeertzjq commented on src/nvim/normal.c:
```diff
@@ -2145,23 +2147,39 @@ void do_check_scrollbind(bool check)
 
+static int get_scrollbind_amount(win_T *ref_wp, int ref_old_topline, int topfill_diff)
+{
+  int ref_orig_topline = ref_wp->w_topline;
```
Hmm, I meant it's not clear that "orig" is newer than old. There is still an "orig" here.

lewis6991 replied:
Sorry, that's meant to say `ref_topline`.

zeertzjq reviewed (commented):
The same change may also be needed in `ex_syncbind()`.

lewis6991 commented:
> The same change may also be needed in `ex_syncbind()`.

I don't think it does. It might need some adjustment to account for filler, but I don't know exactly what.

lewis6991 commented:
Ok, I've had a go. I think this is right.

zeertzjq commented on src/nvim/ex_docmd.c:
```diff
@@ -5593,34 +5594,29 @@ static void ex_swapname(exarg_T *eap)
-  // determine max topline
+  // determine max (virtual) topline
   if (curwin->w_p_scb) {
     topline = curwin->w_topline;
```
You need to use `plines_m_win_fill(wp, 1, wp->w_topline)` here too.

zeertzjq commented on src/nvim/plines.c:
```diff
@@ -907,6 +907,19 @@ int plines_m_win(win_T *wp, linenr_T first, linenr_T last, int max)
+/// Mainly used for calculating scrolling offsets.
+int plines_m_win_fill(win_T *wp, linenr_T first, linenr_T last)
+{
+  int count = 0;
```
Fast path for the common case
```suggestion
  if (!win_may_fill(wp)) {
    return last - first + 1;
  }
  int count = 0;
```

lewis6991 commented on src/nvim/normal.c:
```diff
@@ -2173,35 +2189,33 @@ void check_scrollbind(linenr_T topline_diff, int leftcol_diff)
         } else {
           scrolldown(curwin, -y, false);
         }
+        curwin->w_scbind_pos = curwin->w_topline;
```
Thanks, fixed with testcases.

zeertzjq reviewed (commented):
Some other variables may also need changing "topline" to "vtopline" to avoid confusion.

Also, `w_scbind_pos` is used inconsistently. When passed to `get_scrollbind_amount()` it is used as a physical line number, but then it gets a virtual line count added to it.

zeertzjq commented on src/nvim/ex_docmd.c:
```diff
@@ -5593,45 +5594,43 @@ static void ex_swapname(exarg_T *eap)
     }
   }
+
+  linenr_T old_linenr = curwin->w_cursor.lnum;
```
This doesn't seem correct. It should be assigned before the loop above.

lewis6991 commented on src/nvim/normal.c:
```diff
@@ -2096,17 +2096,23 @@ static void display_showcmd(void)
   static buf_T *old_buf = NULL;
   static colnr_T old_leftcol = 0;
 
+  int vtopline = get_vtopline(curwin);
```
Doing this is more expensive than before. I wonder if there is a way to avoid doing this everytime.

lewis6991 replied:
This should be efficient now as `plines_m_win_fill` now invokes just one iterator on the marktree.

zeertzjq commented on src/nvim/normal.c:
```diff
@@ -2106,42 +2106,44 @@ int get_vtopline(win_T *wp)
 void do_check_scrollbind(bool check)
 {
+  if (!check || !curwin->w_p_scb) {
+    return;
```
The static variables still need to be updated.

lewis6991 replied:
Yes, just realised.

lewis6991 commented:
@luukvbaal do you have any idea how we can refactor `decor_virt_lines` so it doesn't consider folding state and have that logic outside? For context I want to change `decor_virt_lines()` to take a `start_row` and `end_row` as opposed to `lnum` to make calculating filler lines more efficient for scrollbind.

luukvbaal commented:
Sorry I haven't had time to look into this really and it looked like you and zeer had it covered. I also don't really use 'scrollbind'/diff mode so I'm not that familiar with how it even is supposed to work.

For my understanding, is it not a bit ambiguous what is a "useful" scroll amount to keep scrollbound windows in sync in the presence of filler lines? At least when it concerns windows with the same buffer but already different toplines, or even different buffers with different number of filler lines? I think the only thing that would make sense is to always scroll the same number of screen lines, regardless of the number of filler lines. Is that what we're trying to achieve here?

> do you have any idea how we can refactor decor_virt_lines so it doesn't consider folding state and have that logic outside?

I could take a look, but why do we want this? It seems to me that even when checking a line range, whether lines in that range are folded and thus showing filler lines or not is still relevant?

lewis6991 commented:
> I think the only thing that would make sense is to always scroll the same number of screen lines, regardless of the number of filler lines. Is that what we're trying to achieve here?

Yep. You can see the issue being fixed here by running `:Gitsigns blame` where the virtual lines are messing up scrollbind.

This change fixes all the logic by changing from physical lines to screen lines. The problem with that is it involves a lot of line and marktree scanning which I'd like to optimise.

> I could take a look, but why do we want this? It seems to me that even when checking a line range, whether lines in that range are folded and thus showing filler lines or not is still relevant?

Yeah maybe. At the moment this change doesn't consider folded lines at all (future improvement?), but for now I want to at least make sure everything is handled consistently. And tbh I'm not fully convinced scrollbind needs to consider folds since all the calculations are for everything above topline. If the user wants scrollbind with folds, then they just need to ensure the folds are consistent in each window and then it should work regardless.

At the very least I want to pass a range into `decor_virt_lines()` as opposed to calling it for each individual line.

luukvbaal commented:
> At the very least I want to pass a range into decor_virt_lines() as opposed to calling it for each individual line.

That seems fine to me. But rather than refactoring to handle the folds elsewhere, I think it should then handle folds for the entire range. `decor_virt_lines()` should still return what it does currently, even if undocumented. I.e. the number of "active" virtual lines. When checking a range, virtual lines will be active depending on whether it's above/below line is folded. I can't imagine how glancing over that fact for scrollbind would make sense, even if it only checks lines above the current topline?

lewis6991 commented:
> I can't imagine how glancing over that fact for scrollbind would make sense, even if it only checks lines above the current topline?

No where else in scrollbind does it take into account folded lines, so considering it for only virtual lines (always) will lead to edge case bugs in addition to being less efficient.

I also found the fold logic in `decor_virt_lines()` confusing. The code that needs to know about active virtual lines for a specific line number should just call `decor_virt_lines()` with the appropriate range. I don't understand the discontinuity with `row`, `start_row` and `end_row`.

lewis6991 commented:
> I don't think scrollbind needs to take into account folded lines either, at least not the hidden/shown height difference of the hidden buffer lines. But surely a function that returns the number of virtual lines should return the number of active virtual lines, not also include those that are hidden for whatever reason (in this case because they are attached to a closed fold).

No. Whether virtual lines are active or not is a distinction which different callers may or may not care about. win_line wants active, scrollbind does not. And quite frankly should be done at a higher level and not directly in the decor functions.

> Yes, the decor_virt_lines arguments are (somewhat needlessly) confusing for performance reasons. Because it is called from win_line() where has_fold is already known, and lines is populated for further use. But also called from other places where that doesn't hold.

Honestly, seems like a micro-op to me. If saving a single call to `hasFolding` is so important, you should do it for the other line too in a function that's specialised specifically for win_line in drawline.c. I expect in practice this optimisation makes a negligible impact. As opposed to scrollbind that needs to scan everything above the topline on many normal mode operations.

> If you are convinced that we need a function that returns the number of virtual lines for a certain range, regardless of whether they are folded or not, perhaps that should just be a separate function.

Ideally we have one general function, and layer other functions on top for more specialised use cases. If win_line is already passing a pointer for the virt_lines, then doesn't it have all the information it needs to figure out the active virtual lines?

Regardless, has_fold is too weird and should go/(be done differently), and whether active/all virtual lines are needed could be an option.

> Updating the existing function to properly handle folds for the entire range seems to make more sense to me though.

It only makes sense if it's needed. Would be nice, but not needed here. It could be used to optimise plines.c.

---

One thing that I don't understand about the function is that the folding variables only affect the bounds of the iterator. I tried hard coding the bounds of the iterator to be the largest value they need to be but that didn't work. Surely the bounds just need to be big enough to iterate the lines we care about and the actual selection of the marks is filtered in the loop body? Do you know why that doesn't work? If we can hard code the bounds to be larger, then we can avoid all calls to `hasFolding`?

luukvbaal commented:
> Whether virtual lines are active or not is a distinction which different callers may or may not care about. win_line wants active, scrollbind does not. And quite frankly should be done at a higher level and not directly in the decor functions.

Can you explain why it doesn't? I am missing the reason why scrollbind should count inactive virtual lines. Repeating myself here but in my mind anything plines adjacent counting the number of virtual lines surely only cares about ones that are actually visible. Whether hidden by a closed fold, `invalid` extmark property, or otherwise.

> Honestly, seems like a micro-op to me. If saving a single call to hasFolding is so important, you should do it for the other line too in a function that's specialised specifically for win_line in drawline.c. I expect in practice this optimisation makes a negligible impact.

I'm inclined to agree but you should argue that with @zeertzjq, who added it [here](https://github.com/neovim/neovim/pull/21930#issuecomment-1399469275).

> As opposed to scrollbind that needs to scan everything above the topline on many normal mode operations.

Does it make a non-negligible impact there? Even though that still seems irrelevant to me if it is necessary for correctness.

> Ideally we have one general function, and layer other functions on top for more specialised use cases. If win_line is already passing a pointer for the virt_lines, then doesn't it have all the information it needs to figure out the active virtual lines?

Not currently no, the information of whether it belongs to a "below" or "above" line is lost when returned to win_line(). I suppose it could be passed along and handled back at the call site but I'm not sure that would make it more clear.

> One thing that I don't understand about the function is that the folding variables only affect the bounds of the iterator. I tried hard coding the bounds of the iterator to be the largest value they need to be but that didn't work. Surely the bounds just need to be big enough to iterate the lines we care about and the actual selection of the marks is filtered in the loop body? Do you know why that doesn't work? If we can hard code the bounds to be larger, then we can avoid all calls to hasFolding?

Not quite sure I understand. The only reason we are dealing with any bounds to begin with, as opposed to a single line, is because win_line() draws virtual lines for both the previous and currently drawn line, depending on the `virt_lines_above` property. We are only adjusting the bounds so as to filter out virtual lines belonging to folded lines, so that they won't be drawn. Could be done inside the loop body too yeah, and would have to be when handling a larger range, but adjusting the bounds is more efficient currently. I don't see how that could avoid calling `hasFolding()`, unless you mean just returning all virtual lines, and handling it in `win_line()`, given that the returned lines also store the line they belong to. Which again I'm not sure would be an improvement.

lewis6991 commented:
Ok, I've changed `decor_virt_lines()` with the following:

- It now handles a range.
- Removed `has_fold` argument.
- Added `active_only` argument which controls whether folds are considered or not.
- Encapsulated all the fold related logic into a `virt_line_row_is_active()` predicate which is called in the loop.

For lines that don't have virtual lines this should be more efficient as it should avoid any calls to `hasFolding()`, whereas before it was called at least once for any buffer containing at least one virtual line.

For lines that do have virtual lines, this will be slightly less efficient:
- For lines with a single virtual lines mark, `hasFolding()` is always called twice.
   - Virtual lines are rare enough for this not to be worth optimizing, IMO.
- For lines with multiple virtual lines marks, `hasFolding()` is called 2x the amount of marks.
   - This could be optimized, but having multiple virtual lines marks on a single line is very rare.

lewis6991 commented:
Heh, this is funny.

Given:

```c
static bool virt_line_row_is_active(win_T *wp, int row, bool above)
{
  int draw_row = row + (above ? 0 : 1);
  bool row_below_fold = draw_row > 0 && hasFolding(wp, draw_row, NULL, NULL);
  bool row_has_fold = hasFolding(wp, draw_row + 1, NULL, NULL);

  int active_start_row = row_below_fold ? draw_row : MAX(draw_row - 1, 0);
  int active_end_row = row_has_fold ? draw_row : draw_row + 1;

  return row >= active_start_row && row < active_end_row;
}
```

Specialize for `above`:

```c
static bool virt_line_row_is_active(win_T *wp, int row, bool above)
{
  if (above) {
    bool row_below_fold = row + 1 > 0 && hasFolding(wp, row + 1, NULL, NULL);
    bool row_has_fold = hasFolding(wp, row + 2, NULL, NULL);

    int active_start_row = row_below_fold ? row + 1 : row;
    int active_end_row = row_has_fold ? row + 1 : row + 2;

    return row >= active_start_row && row < active_end_row;
  } else {
    bool row_below_fold = row > 0 && hasFolding(wp, row, NULL, NULL);
    bool row_has_fold = hasFolding(wp, row + 1, NULL, NULL);

    int active_start_row = row_below_fold ? row : MAX(row - 1, 0);
    int active_end_row = row_has_fold ? row : row + 1;

    return row >= active_start_row && row < active_end_row;
  }
}
```

Which reduces to:

```c
static bool virt_line_row_is_active(win_T *wp, int row, bool above)
{
  if (above) {
    bool row_below_fold = hasFolding(wp, row + 1, NULL, NULL);
    return row >= (row_below_fold ? row + 1 : row);
  } else {
    bool row_has_fold = hasFolding(wp, row + 1, NULL, NULL);
    return row < (row_has_fold ? row : row + 1);
  }
}
```

which reduces to:

```c
static bool virt_line_row_is_active(win_T *wp, int row, bool above)
{
  return !hasFolding(wp, row + 1, NULL, NULL);
}
```

So it seems we do not need to know if the virtual line is above or not.

lewis6991 commented:
Any more feedback before I merge?
